### ISSUE TITLE
JSON: account->aibalance->astyle->asprecision is an object in 1.19.1

### ISSUE THREAD

**Issue by real-dam** [simonmichael/hledger#1360]
Hi,

I got a crash report from two MoLe users that seems to be a problem in the JSON API of hledger-web. The `asprecision` element, which usually is a number, is an object, and that breaks the parser. Here's the beginning of the `/accounts` output. The interesting part is at the middle:

```
[
   {
      "aboring" : false,
      "aibalance" : [
         {
            "acommodity" : "INR",
            "aismultiplier" : false,
            "aprice" : null,
            "aquantity" : {
               "decimalMantissa" : 0,
               "decimalPlaces" : 10,
               "floatingPoint" : 0
            },
            "astyle" : {
               "ascommodityside" : "R",
               "ascommodityspaced" : true,
               "asdecimalpoint" : ".",
               "asdigitgroups" : null,
               "asprecision" : {
                  "contents" : 2,
                  "tag" : "Precision"
               }
            }
         }
      ],
      "aname" : "root",
      "aparent_" : "",
      "asubs" : [],
      "asubs_" : [
         "ASSETS",
         "EQUITY",
         "EXPENSES",
         "INCOME",
         "…"
      ]
   }
]
```

hledger-web version is reported to be `1.19.1`. I can't reproduce this with the demo.hledger.org installation that always uses a plain number for `asprecision`. The version I personally use with MoLe is `1.16.2` and that is fine too.

Perhaps there is something in the journal file that triggers the issue, but I don't have access to the raw data.

I hope this rings a bell. I should be able to get more information about the journal content if you tell me what to ask for.

In theory, I could spawn another parser hierarchy that requires that `asprecision` is an object. That would be the way if the change in the API was intentional and is there to stay.

Cheers.

**Comment by Xitian9** [comment]
This is because of a change in the internal representation of `Precision`, implemented in f6fa76bba7530af3be825445a1097ae42498b1cd. The change was used to get rid of magic values for `Precision`, but since it is now a sum type it will be serialised in JSON as an object by default.

An option would be to write a custom JSON serialisation which allows it to be represented by a bare number.

**Comment by real-dam** [comment]
Thanks for the analysis.

Would you say that this was an intentional change in the JSON output and I should adapt the parser?

**Comment by Xitian9** [comment]
It was an intentional change in the internal representation. I wouldn't quite say it was intentional that the JSON output changed, but it was a natural consequence.

I would recommend adapting the parser, unless there's a specific reason to not.

**Comment by real-dam** [comment]
Right.
I tried to adapt the parsing code, but decided to drop the `astyle` from the parser structures instead.
The reason is that making the parser support multiple JSON APIs requires a substantial work that is not worth the effort because the style data is not really used.
The parser library already ignores unknown fields, so now the parser works with both APIs, because they differ in a part that is ignored.

Thank you for your help. I am closing this issue, since there is nothing that needs to be done in hledger.

**Comment by simonmichael** [comment]
Sorry about that @real-dam, we were not aware. We must add a test to catch such JSON changes. 

To be clear, in hledger <1.19 asprecision was `Word8`, and since 1.19 it is `Precision Word8 | NaturalPrecision`, which I assume means the JSON will be either what you saw:
```json
               "asprecision" : {
                  "contents" : 2,
                  "tag" : "Precision"
               }
```
or:
```json
               "asprecision" : {
                  "tag" : "NaturalPrecision"
               }
```

**Comment by Xitian9** [comment]
Should we restore the previous JSON representation by storing `Precision n` as just the integer n, and `NaturalPrecision` as a magic value like 256, outside the range of `Word8`?

**Comment by simonmichael** [comment]
Magic values are not ideal, but yes maybe so if this rises up again as a need. (-1 sounds better, in case the 255 limit increases.)

**Comment by simonmichael** [comment]
Oh, now I see #1510. I like it, a JSON null for natural precision. @real-dam how does it sound for you ? You'd have to update your client again I think.

**Comment by real-dam** [comment]
MoLe can adopt.
If the change is made available on demo.hledger.org, that would give me some time to prepare MoLe before the new protocol is released.

**Comment by simonmichael** [comment]
@real-dam, will do. And I'm assuming the change is a good one for you and other clients in the long run. I'll merge soon.

**Comment by simonmichael** [comment]
Merged #1510, updated https://demo.hledger.org. And now we have at least one test of the JSON output, so this issue can be closed (thanks @Xitian9).
